transport/x/grpc/inbound.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package grpc
+
+import (
+	"errors"
+	"net"
+	"sync"
+
+	"go.uber.org/yarpc/api/transport"
+	internalsync "go.uber.org/yarpc/internal/sync"
+
+	"github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc"
+	"google.golang.org/grpc"
+)
+
+var (
+	errRouterNotSet          = errors.New("router not set")
+	errRouterHasNoProcedures = errors.New("router has no procedures")
+
+	_ transport.Inbound = (*Inbound)(nil)
+)
+
+// Inbound is a grpc transport.Inbound.
+type Inbound struct {
+	once           internalsync.LifecycleOnce
+	lock           sync.Mutex
+	listener       net.Listener
+	inboundOptions *inboundOptions
+	router         transport.Router
+	server         *grpc.Server
+}
+
+// NewInbound returns a new Inbound for the given listener.
+func NewInbound(listener net.Listener, options ...InboundOption) *Inbound {
```

Comment:
if it's mostly for testing I'm opposed to making it the first entry point into creating an inbound because it complicates our model (if we look at most of the places where we create a GRPC inbound we have a few extra lines to create the net listener). 

What if we have two other inboundOptions:
```
WithListener(net.Listener)
WithAddress(port int)
```
*shrug*